crates/sui-graphql-rpc/src/context_data/db_data_provider.rs

```diff
@@ -506,15 +508,51 @@ impl PgManager {
             .await
             .map_err(|e| Error::Internal(e.to_string()))
     }
+
+    /// Takes a query fragment and a lambda that executes the query
+    /// Spawns a blocking task that determines the cost of the query fragment
+    /// And if within limits, executes the query
+    async fn run_query_async_with_cost<T, Q, EF, E, F>(
+        &self,
+        query: Q,
+        execute_fn: EF,
+    ) -> Result<T, Error>
+    where
+        Q: QueryDsl
+            + RunQueryDsl<Pg>
+            + diesel::query_builder::QueryFragment<diesel::pg::Pg>
+            + Send
+            + 'static,
+        EF: FnOnce(Q) -> F + Send + 'static,
+        F: FnOnce(&mut PgConnection) -> Result<T, E> + Send + 'static,
+        E: From<diesel::result::Error> + std::error::Error + Send + 'static,
+        T: Send + 'static,
+    {
+        let max_db_query_cost = self.limits.max_db_query_cost;
```

Comment:
This limit applies to all queries that are run as part of a single request, but currently it looks like we're judging the cost of each transaction individually.

Reply:
This was a q I had posed to @oxade , and I think I was convinced by the explanation (unless I misunderstood it) that the node and complexity limiting already cover the whole request, so the query cost limiting is just to make sure that each individual db operation is not overly complex

Reply:
Ah okay, my recollection of past discussions was that this step was also intended to cover the whole request, so that each successive cost check (node + depth, explain, timeout) is successively more rigorous, because otherwise, a bad actor can take advantage of a discrepancy between our least and most expensive node to compute:

If we set a local explain limit (like in this PR), it has to be generous enough to run a sensible query involving our cheaper nodes, but then the bad actor can fill up a query with really expensive nodes to query (like the analytics queries for explorer, or the APY query), and overload the system.  If we set a global limit, we don't have that problem.